### PR TITLE
add base profile

### PR DESCRIPTION
**Pull request by cqc-melf**
# Description

Add support for the base profile

# Related issues

Please mention any github issues addressed by this PR.

# Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: tests/qir/test_pytket_qir_20-QIRProfile.BASE.ll [hunk 1]
```diff
@@ -0,0 +1,71 @@
+; ModuleID = 'test_pytket_qir_20'
+source_filename = "test_pytket_qir_20"
+
+%Qubit = type opaque
+%Result = type opaque
+
+@0 = internal constant [5 x i8] c"c[0]\00"
```
Comment (cqc-melf) [inline tests/qir/test_pytket_qir_20-QIRProfile.BASE.ll:7]: We need to record the results individually, there might be device specific options to do this is a better way. I think this is the best way to do it compatible to the base profile.

File: pytket/qir/conversion/baseprofileqirgenerator.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import pyqir
+from pyqir import Value
+
+from pytket import Bit, Circuit, Qubit, wasm  # type: ignore
```
Comment (cqc-alec) [inline pytket/qir/conversion/baseprofileqirgenerator.py]: Prefer `from pytket.wasm import WasmFileHandler` and use that directly; then we may not need this "type: ignore"?

File: pytket/qir/conversion/baseprofileqirgenerator.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        raise ValueError("Classical REG Ops Not available in base profile")
+
+    def set_ssa_vars(self, reg_name: str, ssa_i64: Value, trunc: bool) -> None:
+        raise ValueError("Classical REG Ops Not available in base profile")
+
+    def _reg2ssa_var(self, bit_reg: BitRegister) -> Value:
+        raise ValueError("Classical REG Ops Not available in base profile")
+
+    def conv_conditional(self, command: Command, op: Conditional) -> None:
+        raise ValueError("Conditional NOT available in base profile")
```
Comment (cqc-alec) [inline pytket/qir/conversion/baseprofileqirgenerator.py]:
```suggestion
        raise ValueError("Conditional not available in base profile")
```

File: pytket/qir/conversion/baseprofileqirgenerator.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                pyqir.Type.void(self.module.module.context),
+                [
+                    pyqir.result_type(self.module.module.context),
+                    pyqir.PointerType(pyqir.IntType(self.module.module.context, 8)),
+                ],
+            ),
+        )
+
+    def _get_bit_from_creg(self, creg: str, index: int) -> Value:
+        raise ValueError("Classical REG Ops Not available in base profile")
```
Comment (cqc-alec) [inline pytket/qir/conversion/baseprofileqirgenerator.py]:
```suggestion
        raise ValueError("Classical register operations not available in base profile")
```
(similarly below)

File: pytket/qir/conversion/baseprofileqirgenerator.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    def _reg2ssa_var(self, bit_reg: BitRegister) -> Value:
+        raise ValueError("Classical REG Ops Not available in base profile")
+
+    def conv_conditional(self, command: Command, op: Conditional) -> None:
+        raise ValueError("Conditional NOT available in base profile")
+
+    def conv_measure(self, bits: list[Bit], qubits: list[Qubit]) -> None:
+
+        qubit_index = qubits[0].index[0]
```
Comment (cqc-alec) [inline pytket/qir/conversion/baseprofileqirgenerator.py:108]: Should probably (a) check that `bits` and `qubits` are the same length and (b) handle the case where that length is 0.

Comment (cqc-melf) [reply pytket/qir/conversion/baseprofileqirgenerator.py:108]: They should always have exactly one entry, I have added an assert. See [40b2085](https://github.com/CQCL/pytket-qir/pull/177/commits/40b2085eabba5552288d5969a799063d171e15eb)

File: pytket/qir/conversion/baseprofileqirgenerator.py (hunk 2 above)
Comment (cqc-melf) [reply pytket/qir/conversion/baseprofileqirgenerator.py]: I have changed that, see [195660a](https://github.com/CQCL/pytket-qir/pull/177/commits/195660a65126a4787976751d51e98bc49fe8c481)
The changes needed to make mypy happy ended up a bit bigger.

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**